app/javascript/components/student/Editor.tsx

```diff
@@ -24,32 +66,61 @@ export function Editor({
   endpoint: string
   timeout?: number
 }) {
-  const [submission, setSubmission] = useState<Submission>()
-  const controllerRef = useRef(new AbortController())
+  const [{ submission, status }, dispatch] = useReducer(reducer, {
+    status: undefined,
+    submission: undefined,
+  })
+  const controllerRef = useRef<AbortController | undefined>(
+    new AbortController()
+  )
+  const abort = useCallback(() => {
+    controllerRef.current?.abort()
+    controllerRef.current = undefined
+  }, [controllerRef])
+  const submit = useCallback(
+    (code: string) => {
+      abort()
+      controllerRef.current = new AbortController()
+
+      dispatch({ type: EditorStatus.SUBMITTING })
+
+      fetchJSON(endpoint, {
+        method: 'POST',
+        signal: controllerRef.current.signal,
+        body: JSON.stringify({ files: { file: code } }),
+      })
+        .then((json: any) => {
+          dispatch({
+            type: EditorStatus.SUBMITTED,
+            payload: { submission: typecheck<Submission>(json, 'submission') },
+          })
+        })
+        .catch((err) => {
+          if (err.name === 'AbortError' && controllerRef.current) {
```

Comment:
Why must the controllerRef be checked? What's the thinking here? 

If the idea is "well on unmount we abort, and then therefore we unset it", it's probably nicer for readers to have a _dedicated_ ref that tracks this. If we now refactor `abort`, this will break.

Reply:
You shouldn't dispatch the `cancelled` event if there is an existing fetch going on. This happens when submit is clicked twice in a row.